samples/zigbee/light_switch/src/nus_cmd.c

```diff
@@ -0,0 +1,219 @@
+ */
+#include <zephyr.h>
+#include <bluetooth/gatt.h>
+#include <bluetooth/hci.h>
+#include <bluetooth/uuid.h>
+#include <bluetooth/services/nus.h>
+#include <logging/log.h>
+#include <settings/settings.h>
+
+#include "nus_cmd.h"
+
+LOG_MODULE_REGISTER(nus_cmd);
+
+#define DEVICE_NAME        CONFIG_BT_DEVICE_NAME
+#define DEVICE_NAME_LEN    (sizeof(DEVICE_NAME) - 1)
+
+static void connected(struct bt_conn *conn, u8_t err);
+static void disconnected(struct bt_conn *conn, u8_t reason);
+static void auth_passkey_display(struct bt_conn *conn, unsigned int passkey);
+static void auth_cancel(struct bt_conn *conn);
+static void pairing_confirm(struct bt_conn *conn);
+static void pairing_complete(struct bt_conn *conn, bool bonded);
+static void pairing_failed(struct bt_conn *conn, enum bt_security_err reason);
+static void __attribute__((unused))
+security_changed(struct bt_conn *conn, bt_security_t level,
+		 enum bt_security_err err);
+
+static struct bt_conn_auth_cb conn_auth_callbacks = {
+	.passkey_display = auth_passkey_display,
+	.cancel = auth_cancel,
+	.pairing_confirm = pairing_confirm,
+	.pairing_complete = pairing_complete,
+	.pairing_failed = pairing_failed
+};
+
+static struct bt_conn_cb conn_callbacks = {
+	.connected = connected,
+	.disconnected = disconnected,
+	COND_CODE_1(CONFIG_BT_SMP, (.security_changed = security_changed), ())
+};
+
+static const struct bt_data ad[] = {
+	BT_DATA_BYTES(BT_DATA_FLAGS, (BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR)),
+	BT_DATA(BT_DATA_NAME_COMPLETE, DEVICE_NAME, DEVICE_NAME_LEN),
+};
+
+static const struct bt_data sd[] = {
+	BT_DATA_BYTES(BT_DATA_UUID128_ALL, NUS_UUID_SERVICE),
+};
+
+static struct k_work on_connect_work;
+static struct k_work on_disconnect_work;
+static struct bt_conn *current_conn;
+static nus_entry_t *nus_commands;
+
+static void connected(struct bt_conn *conn, u8_t err)
+{
+	if (err) {
+		LOG_ERR("Connection failed (err %u)", err);
+		return;
+	}
+
+	LOG_INF("Connected");
+	current_conn = bt_conn_ref(conn);
+
+	k_work_submit(&on_connect_work);
+}
+
+static void disconnected(struct bt_conn *conn, u8_t reason)
+{
+	LOG_INF("Disconnected (reason %u)", reason);
+
+	if (current_conn) {
+		bt_conn_unref(current_conn);
+		current_conn = NULL;
+
+		k_work_submit(&on_disconnect_work);
+	}
+}
+
+static char *ble_addr(struct bt_conn *conn)
+{
+	static char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	return addr;
+}
+
+static void __attribute__((unused))
+security_changed(struct bt_conn *conn, bt_security_t level,
+		 enum bt_security_err err)
+{
+	char *addr = ble_addr(conn);
+
+	if (!err) {
+		LOG_INF("Security changed: %s level %u", log_strdup(addr),
+			level);
+	} else {
+		LOG_INF("Security failed: %s level %u err %d", log_strdup(addr),
+			level, err);
+	}
+}
+
+static void auth_passkey_display(struct bt_conn *conn, unsigned int passkey)
+{
+	char *addr = ble_addr(conn);
+
+	/* Logging this in the production code is unsafe */
```

Comment:
you should justify why.

Reply:
@joerchan I'm not the original author of this. Indeed, if you can read log output the you in possession of the device and it doesn't seem to be an issue. I'll let @lmaciejonczyk comment.

Reply:
I suppose @MarekPorwisz wants to answer this, as it was added  on his request.

Reply:
IMO: We do not know what is the end user use-case. passkey is a sensitive data and it should be emphasized it is unsafe to share/log it. The logs are not necessarily stored on the device, there are dozens of centralized logging features out there. In case of failures logs are sent via emails or other unsafe and sometimes encrypted channels. Pass-codes, passphrases, passkeys, passwords shall never be written into logs. Originally I asked to remove it from logs but was convinced it may be useful for the user to see that. So I asked requested adding a comment that this is unsafe for production code.

Reply:
@MarekPorwisz In general yes. However, the question here is whether you can actually do anything useful with this passkey after it has been used.

Reply:
I'm not that familiar with Bluetooth so I can't say if the passkey is still valid at the moment of printing or for how long more will it be. I just see passkey being printed here and that's a yellow light for me there might be something wrong with data security here. If the passkey is no longer valid it should be stated in the comment that logging passkey here is not a vulnerability because ....

Reply:
The passkey is used as a one-time authentication. I think you can remove this comment.